packages/grafana-ui/src/components/Combobox/Combobox.tsx

```diff
@@ -144,34 +158,65 @@ export const Combobox = <T extends string | number>({
     defaultHighlightedIndex: selectedItemIndex ?? 0,
     scrollIntoView: () => {},
     onInputValueChange: ({ inputValue }) => {
-      const filteredItems = options.filter(itemFilter(inputValue));
-      if (createCustomValue && inputValue && filteredItems.findIndex((opt) => opt.label === inputValue) === -1) {
-        const customValueOption: ComboboxOption<T> = {
-          label: inputValue,
-          // @ts-ignore Type casting needed to make this work when T is a number
-          value: inputValue as unknown as T,
-          description: t('combobox.custom-value.create', 'Create custom value'),
-        };
-
-        setItems([...filteredItems, customValueOption]);
+      const customValueOption =
+        createCustomValue &&
+        inputValue &&
+        items.findIndex((opt) => opt.label === inputValue || opt.value === inputValue) === -1
+          ? {
+              // Type casting needed to make this work when T is a number
+              value: inputValue as unknown as T,
+              description: t('combobox.custom-value.create', 'Create custom value'),
+            }
+          : null;
+
+      if (isAsync) {
+        if (customValueOption) {
+          setItems([customValueOption]);
+        }
+        setAsyncLoading(true);
+        loadOptions(inputValue)
+          .then((opts) => {
+            setItems(customValueOption ? [customValueOption, ...opts] : opts);
+            setAsyncLoading(false);
+          })
+          .catch((err) => {
+            if (!(err instanceof StaleResultError)) {
+              // TODO: handle error
+              setAsyncLoading(false);
+            }
+          });
+
         return;
-      } else {
-        setItems(filteredItems);
       }
+
+      const filteredItems = options.filter(itemFilter(inputValue));
+
+      setItems(customValueOption ? [customValueOption, ...filteredItems] : filteredItems);
     },
+
     onIsOpenChange: ({ isOpen }) => {
       // Default to displaying all values when opening
-      if (isOpen) {
+      if (isOpen && !isAsync) {
         setItems(options);
         return;
       }
+
+      if (isOpen && isAsync) {
+        setAsyncLoading(true);
+        loadOptions('').then((options) => {
+          setItems(options);
+          setAsyncLoading(false);
+        });
```

Comment:
will errors here just be silently lost? 🤔

Reply:
yup! (for now) ✨